packages/web-components/src/theme/set-theme.ts

```diff
@@ -3,34 +3,53 @@ import * as tokens from './design-tokens.js';
 
 const tokenNames = Object.keys(tokens) as (keyof Theme)[];
 
+const SUPPORTS_REGISTER_PROPERTY = 'registerProperty' in CSS;
+const SUPPORTS_ADOPTED_STYLE_SHEETS = 'adoptedStyleSheets' in document;
+const themeStyleSheet = new CSSStyleSheet();
+const themeStyleTextMap = new Map<Theme, string>();
+
 /**
  * Sets the theme tokens on defaultNode.
  * @param theme - Flat object of theme token values.
  * @internal
  */
 export const setTheme = (theme: Theme) => {
-  for (const t of tokenNames) {
-    let registered = false;
-
-    if ('registerProperty' in CSS) {
-      try {
-        CSS.registerProperty({
-          name: `--${t}`,
-          inherits: true,
-          initialValue: theme[t] as string,
-        });
-        registered = true;
-      } catch {
-        // Do nothing.
+  // Fallback to setting token custom properties on `<body>` element’s `style`
+  // attribute, with the assumption that if a user agent supports
+  // `document.adoptedStyleSheets`, it’d also support `CSS.registerProperty()`.
+  if (!SUPPORTS_ADOPTED_STYLE_SHEETS) {
+    setThemeFor(document.body, theme);
+    return;
+  }
+
+  if (!themeStyleTextMap.has(theme)) {
+    const tokenDeclarations: string[] = [];
+
+    for (const t of tokenNames) {
+      if (SUPPORTS_REGISTER_PROPERTY) {
```

Comment:
Could this check be done once, outside of the for-loop?

Reply:
Yes! That was one of the refactor goals and somehow I let it slip. Thanks for catching it!

Reply:
Looked at this again, and remembered that I can't do the check outside of the loop without looping it twice since we do need to build the `tokenDeclarations` array for cache, which is why I stored whether the support is there in a variable (`SUPPORTS_REGISTER_PROPERTY`).